keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1054 @@
+the resource on a node). During scheduling of a pod, kube-scheduler and the DRA
+driver controller(s) that handle claims used by the pod communicate through the
+apiserver by updating the PodSchedulingContext object, ultimately leading to
+allocation of all pending claims and the pod being scheduled onto a node.
+
+This approach poses a problem for [Cluster
+Autoscaler](https://github.com/kubernetes/autoscaler) (CA) because it cannot
+simulate the effect of allocating or deallocating claims. "Semantic parameters"
+is an extension of DRA that addresses this by making claim parameters less
+opaque. "Semantic models" define how resource capacity per node can be published
+and the claim parameters that allocate some of that capacity for a claim.
+
+This KEP describes the core changes needed for this. Specific semantic models
+get defined in separate KEPs. The types defined by these models become
+part of the resource.k8s.io API. Therefore developers who want to extend
+a scheduler with some additional model have to modify the apiserver.
+
+When opting into using semantic parameters for their DRA driver, driver
+developers no longer need to provide a control plane controller which
+handles claim allocation. Instead, they need to provide a CRD controller
+which translates from their custom claim parameters CRD into the parameter
+object understood by Kubernetes.
+
+Scheduling performance is expected to become better compared to using the
+PodSchedulingContext. However, supporting CA is the main reason for this KEP.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- Enable cluster autoscaling when pods use resource claims, with correct
+  decisions and changing the cluster size by more than one node at a time.
+
+- Support node-local resources. Adding or removing nodes has no effect
+  on network-attached resources and therefore CA does not need to (and cannot)
+  simulate them.
+
+- Allow DRA driver developers to provide a user experience that is similar to
+  the one possible without semantic parameters. Ideally, users should not notice
+  at all that a driver is using semantic parameters under the hood.
+
+- Keep the core DRA functionality from [KEP
+  #3063](https://github.com/kubernetes/enhancements/issues/3063) working as
+  defined there for Kubernetes 1.29. Cluster autoscaling is not always needed
+  and in-tree numeric model(s) are not expected to cover all use case, so the
+  flexibility offered by opaque parameters remains relevant.
+
+### Non-Goals
+
+- Support semantic class parameters. They remain opaque and can be used to
```

Comment:
I am not sure what "semantic class parameters" are. Are you trying to say that it is a non-goal to eliminate the need for a claim parameter CR? It is something I have wondered - are there use cases where the semantic selection and capacity request API would be sufficient (e.g., where no real configuration is needed, just selection and allocation)? I don't have an answer to that, but it would be a shame if the API forced vendors to create a CRD and a controller that basically just did a "copy from CR to standard API".

Reply:
The core DRA has claim and class parameters: both ResourceClaim and ResourceClass link to CRDs. What I am saying here is that with semantic parameters, the class parameters have to be limited to configuration because this KEP is not trying to do anything with them.

This is a simplification. We could also extend the KEP later to translate class parameters similar to how we do it now for claim parameters.

> It is something I have wondered - are there use cases where the semantic selection and capacity request API would be sufficient (e.g., where no real configuration is needed, just selection and allocation)?

I was wondering the same. This could be done by letting users create ResourceClaimParameter objects themselves and referring to those in their ResourceClaim. We loose validation, though. I don't know useful this would be. Shall I list it under "alternatives" for future reference?

Reply:
> It is something I have wondered - are there use cases where the semantic selection and capacity request API would be sufficient (e.g., where no real configuration is needed, just selection and allocation)? I don't have an answer to that, but it would be a shame if the API forced vendors to create a CRD and a controller that basically just did a "copy from CR to standard API".

We could define a "standard" `ClaimParamaters` object that the scheduler knows how to interpret directly. If that is referenced by a claim, then the scheduler interprets it directly -- otherwise is delegates this to an external controller.

Reply:
This is exactly what I was proposing. Perhaps you hadn't seen my comment yet?

We already define a standard `ResourceClaimParamaters` type. It's really just a question whether we allow users to reference it directly and whether that would be useful.

Reply:
You’re right, I basically repeat what you said in your last paragraph. I must have missed that somehow — good to know we are in the same page though.

Reply:
I don't love having an indirection for the standard claim parameters. Seems like if it's standard, it can be combined into the ResourceClaim directly. But then you either need to allow that OR a reference to a vendor specific one, or (better, in my mind) you offer both. What would that look like? The ResourceClaim would allow you to specify the selection expression across attributes (whether they are vendor specific or not), as well as allocations from allocatable resources. This effectively like device plugin today. Any vendor-specific "config" would go into the vendor-specific CR. You could ALSO use your controller to process the vendor-specific CR and add ADDITIONAL contraints and ADDITIONAL allocations. 

The indirection is something we *have* to do because of how our API works. It's not better for the user. So once it is a common K8s API, the indirection is not needed. But we do still need it for vendor specific things. Allowing both allows the API to be simpler for simpler cases, but still work for complex cases.

Reply:
Unfortunately the alternatives for getting configuration parameters for class and claim down into the kubelet are all worse.

When the DRA kubelet plugin itself talks to the API server. Then we have issues with pods not starting (parameter object deleted), starting not as intended (parameter object modified after allocation), and less security (permission to read all objects, not just those relevant for the node - probably not that relevant here).

Would it be acceptable to have a RawExtension in the `ResourceClaimParameters` object? Then the driver can can store whatever versioned object it wants during the translation from claim parameters.

We would need a translation from "class parameters" as well. Because we don't want normal users to write those, it would have to be a different ResourceClassParameters resource.

Reply:
> Would it be acceptable to have a RawExtension in the ResourceClaimParameters object?

Yes. But then maybe it should be copied by a k8s controller into the Claim. This is similar to how we deal with Priority: The user sets a priorityClass, but the priority value is copied into the podSpec (I think this happens in the apiserver, but don't quote me on that).

Then kube-scheduler and kubelet only have to look at one object, avoiding retries